### PR TITLE
[VFX] Add support material queue offset

### PR DESCRIPTION
**Pull request by PaulDemeulenaere**
### Purpose of this PR
Add support of material queue offset :
image
This feature is mainly added for URP (because renderer priority is only available on HDRP) but it's compatible with any SRP.

_This solution is a V2 : we change the "slot intput" approach by a "settings", see this [conversation](https://unity.slack.com/archives/G1BTWN88Z/p1615798166121200) to be consistent with the incoming ShaderGraph integration refactor (see https://github.com/Unity-Technologies/Graphics/pull/3467)._

Additions visible to users :
- Add material offset setting in inspector of every output.
- Documented [here](https://github.com/Unity-Technologies/Graphics/blob/2e9b271789aac2eb07981fe0b590e111603ccbda/com.unity.visualeffectgraph/Documentation~/Context-OutputSharedSettings.md)

See a sample of usage here :
_material_offset_sample

**HDRP** : This field is only available for transparent output.
_material_offset_filter

From code perspective, in VFXSRPSubOutput  :
- Added a caps [supporsMaterialOffset](https://github.com/Unity-Technologies/Graphics/pull/3598/files#diff-009906407161538fd811cc6fcf58d7cabe4117a6a0ebb50ba601d385f2bb61f3R28) : true by default (HDRP returns false for opaque)
Note : URP exactly corresponds to builtin renderqueue offset, we don't need to declare a specific `VFXSRPSubOutput`, see [HDRP specific implementation.](https://github.com/Unity-Technologies/Graphics/pull/3598/files#diff-fb69c61dc9ec41e271c3962fe2b2537c2347544f6cbb764a415c674edf2856d8R137)

---
### Testing status
Yamato & Manual
image
The green particles are actually in front of red ones.
https://user-images.githubusercontent.com/22494706/111310842-d0514680-865d-11eb-847f-648e9728fdad.mp4

---
### Comments to reviewers
Related to this [JIRA](https://jira.unity3d.com/browse/GVFXG-107)

### CODE REVIEW

**Comment by PaulDemeulenaere** [comment]
Moved the JIRA to blocked awaiting for C++ (see [here](https://jira.unity3d.com/browse/GVFXG-107?focusedCommentId=1980831&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1980831))
⏳

File: com.unity.visualeffectgraph/Documentation~/Context-OutputSharedSettings.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ All outputs share these settings and property ports.
...
 | **Use Alpha Clipping**      | Bool     | Indicates whether the system uses alpha clipping and discards pixels under a certain alpha threshold. If you enable this property, use the **Alpha Threshold** input to set the alpha value below which the system discards pixels. |
+| **Material Offset**         | Int      | **(Inspector)** Apply an offset the render ordering. This setting corresponds to the Material [Sorting Priority](https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@latest?subfolder=/manual/Renderer-And-Material-Priority.html) with HDRP and Material [Priority](https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@latest?subfolder=/manual/simple-lit-shader.html) with URP and is an equivalent of modify the [Material.renderQueue](https://docs.unity3d.com/ScriptReference/Material-renderQueue.html) property. This value can be negative or positive, a clamp between minimal and maximal range will be applied. The internal range is specific by SRP : [-50;50] with URP, [-100;100] with HDRP. |
```
Comment (JordanL8) [inline com.unity.visualeffectgraph/Documentation~/Context-OutputSharedSettings.md:11]: ``` **(Inspector)** An offset Unity applies to the rendering order. This setting corresponds to the Material [Sorting Priority](https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@latest?subfolder=/manual/Renderer-And-Material-Priority.html) in HDRP and Material [Priority](https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@latest?subfolder=/manual/simple-lit-shader.html) in URP. It is equivalent to modifying the [Material.renderQueue](https://docs.unity3d.com/ScriptReference/Material-renderQueue.html) property. This value can be negative or positive. URP clamps this value between -50 and 50, and HDRP clamps this value between -100 and 100. ```

Comment (JordanL8) [inline com.unity.visualeffectgraph/Documentation~/Context-OutputSharedSettings.md:11]: ``` **(Inspector)** An offset Unity applies to the rendering order. This setting corresponds to the Material [Sorting Priority](https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@latest?subfolder=/manual/Renderer-And-Material-Priority.html) in HDRP and Material [Priority](https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@latest?subfolder=/manual/simple-lit-shader.html) in URP. It is equivalent to modifying the [Material.renderQueue](https://docs.unity3d.com/ScriptReference/Material-renderQueue.html) property. This value can be negative or positive. URP clamps this value between -50 and 50, and HDRP clamps this value between -100 and 100. ```

**Review by JordanL8 — CHANGES_REQUESTED** [review]

**Comment by sebastienlagarde** [comment]
@PaulDemeulenaere  hey, what is the status on this one?

**Comment by PaulDemeulenaere** [comment]
> @PaulDemeulenaere hey, what is the status on this one?

I'm waiting for trunk fix about build to double check the status of newly introduced screenshot before mark it as "ready to review".
@vlad-unity is going to tackle this verification.
I will have to apply documentation update as well.

**Review by VladNeykov — COMMENTED** [review]
Hi, it's looking good, FTP passes, no issues found.
I've jotted down a couple of UX questions on the [test doc](https://docs.google.com/document/d/12ZRCTFbeWNWGbJEbHpixNGqf9VDeLJvpwuqx8R8IHtU/edit#).

File: com.unity.visualeffectgraph/Documentation~/Context-OutputSharedSettings.md (hunk 1 above)
Comment (PaulDemeulenaere) [reply com.unity.visualeffectgraph/Documentation~/Context-OutputSharedSettings.md]: Fixed at https://github.com/Unity-Technologies/Graphics/pull/3598/commits/217bfbf6d82ce287cc997929e24fb8e20cec513d

**Comment by PaulDemeulenaere** [comment]
@vlad-unity 

1) Add delayed on material offset at https://github.com/Unity-Technologies/Graphics/pull/3598/commits/70597fac82a211c6f3dea7b928835371a58a034d
_delayed_field

2) About `Material Offset for Additive`, HDRP doesn't filter the sorting priority in case of additive :
_additive_sorting

**Review by JordanL8 — APPROVED** [review]

**Review by VladNeykov — APPROVED** [review]
LGTM

**Review by julienf-unity — APPROVED** [review]

**Comment by PaulDemeulenaere** [comment]
Thanks for the review, I'm updating internal screenshot with this [yamato job](https://yamato.cds.internal.unity3d.com/jobs/339-ScriptableRenderPipelinePrivate/tree/vfx%252Ffeature%252Fmaterial-sorting-support) with this [PR](https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/pull/153)
Re-running public VFX verification before merge ⏳

**Comment by PaulDemeulenaere** [comment]
The failure about `HDRP_Hybrid on Win_DX11_playmode_mono_Linear` doesn't seem related to this change.
Ready to be merge, the internal PR has been merged [here](https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/commit/d2d5704209921506ab4df9c9a121cb4f145d9a34)

**Pull request merged**